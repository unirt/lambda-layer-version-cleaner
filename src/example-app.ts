import * as cdk from 'aws-cdk-lib';
import * as events from 'aws-cdk-lib/aws-events';
import { LambdaLayerVersionCleaner } from './index';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'ExampleStack');

new LambdaLayerVersionCleaner(stack, 'LambdaLayerVersionCleaner', {
  retainVersions: '5',
  layerCleanerSchedule: events.Schedule.rate(cdk.Duration.days(7)),
});