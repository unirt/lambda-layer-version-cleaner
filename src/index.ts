import * as cdk from 'aws-cdk-lib';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';
import { DeleteOldLambdaLayerVersionsFunction } from './lambda/delete-old-lambda-layer-versions-function';

/**
 * Properties for `LambdaLayerVersionCleaner`
 */
export interface ILambdaLayerVersionCleanerProps {
  /**
   * Number of versions to retain (default is 10)
   * @default "10"
   */
  retainVersions?: string;

  /**
   * Schedule for the function execution (default is once per day)
   * @default events.Schedule.rate(cdk.Duration.days(1))
   */
  layerCleanerSchedule?: events.Schedule;

  /**
   * Maximum allowed runtime for the Lambda function (default is 15 minutes)
   * @default cdk.Duration.minutes(15)
   */
  handlerTimeout?: cdk.Duration;

  /**
   * Amount of memory allocated to the Lambda function (default is 256MB)
   * @default 256
   */
  handlerMemorySize?: number;
}

/**
 * Lambda Layer Version Cleaner Construct
 *
 * This construct creates a Lambda function that deletes old versions of a Lambda Layer. The function is
 * scheduled to run at a regular interval using an EventBridge rule. The number of versions to retain, as
 * well as the function execution schedule, can be customized using the `ILambdaLayerVersionCleanerProps`
 * interface.
 */
export class LambdaLayerVersionCleaner extends Construct {
  readonly handler: lambda.Function;
  readonly rule: events.Rule;

  constructor(scope: Construct, id: string, props?: ILambdaLayerVersionCleanerProps) {
    super(scope, id);

    // Create the Lambda function to delete old layer versions
    this.handler = new DeleteOldLambdaLayerVersionsFunction(this, 'DeleteOldLambdaLayerVersionsFunction', {
      environment: { RETAIN_VERSIONS: props?.retainVersions ?? '10' }, // Set default retainVersions to 10
      timeout: props?.handlerTimeout ?? cdk.Duration.minutes(15), // Set default timeout to 15 minutes
      memorySize: props?.handlerMemorySize ?? 256, // Set default memory size to 256MB
    });

    // Create the LogGroup for the Lambda function
    new logs.LogGroup(this, `${this.handler.node.id}LogGroup`, {
      logGroupName: `/aws/lambda/${this.handler.functionName}`,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // Add the necessary IAM permissions for the Lambda function
    const handlerPolicyStatement = new iam.PolicyStatement({
      actions: ['lambda:ListLayers', 'lambda:ListLayerVersions', 'lambda:DeleteLayerVersion'],
      effect: iam.Effect.ALLOW,
      resources: ['*'],
    });
    this.handler.addToRolePolicy(handlerPolicyStatement);

    // Create the EventBridge rule for the Lambda function execution schedule
    this.rule = new events.Rule(this, 'LayerCleanerScheduleRule', {
      schedule: props?.layerCleanerSchedule ?? events.Schedule.rate(cdk.Duration.days(1)), // Set default schedule to once per day
    });
    this.rule.addTarget(new targets.LambdaFunction(this.handler));

    // Add the necessary permissions for the EventBridge rule to invoke the Lambda function
    targets.addLambdaPermission(this.rule, this.handler);
  }
}