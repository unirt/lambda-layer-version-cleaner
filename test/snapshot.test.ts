import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { LambdaLayerVersionCleaner } from '../src/index';

test('snapshot test', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'ExampleStack');
  new LambdaLayerVersionCleaner(stack, 'LambdaLayerVersionCleaner');
  const template = Template.fromStack(stack).toJSON();

  expect(template).toMatchSnapshot();
});