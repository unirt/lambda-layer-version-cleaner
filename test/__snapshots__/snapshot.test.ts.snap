// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "LambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunction28B1F8EF": Object {
      "DependsOn": Array [
        "LambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunctionServiceRoleDefaultPolicy0ECA84B8",
        "LambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunctionServiceRoleEDFD15B3",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b65454a40357a6da6d370acb3be4ff43b6327e1fd8437a697aac959dd681873f.zip",
        },
        "Description": "src/lambda/delete-old-lambda-layer-versions.lambda.ts",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "RETAIN_VERSIONS": "10",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunctionServiceRoleEDFD15B3",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunctionLogGroup719934B4": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "LambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunction28B1F8EF",
              },
            ],
          ],
        },
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "LambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunctionServiceRoleDefaultPolicy0ECA84B8": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "lambda:ListLayers",
                "lambda:ListLayerVersions",
                "lambda:DeleteLayerVersion",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunctionServiceRoleDefaultPolicy0ECA84B8",
        "Roles": Array [
          Object {
            "Ref": "LambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunctionServiceRoleEDFD15B3",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunctionServiceRoleEDFD15B3": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaLayerVersionCleanerLayerCleanerScheduleRule19B25384": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(1 day)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "LambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunction28B1F8EF",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "LambdaLayerVersionCleanerLayerCleanerScheduleRuleAllowEventRuleExampleStackLambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunction9C0A7640D3B4CCBB": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "LambdaLayerVersionCleanerDeleteOldLambdaLayerVersionsFunction28B1F8EF",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "LambdaLayerVersionCleanerLayerCleanerScheduleRule19B25384",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
